### ISSUE TITLE
How to combine augmented-ui and glow?

### ISSUE THREAD

**Issue by maxkahnt** [propjockey/augmented-ui#11]
Hey, I appreciate the toolbox you provide and I am looking to incorporate it in a recent project.

So far I failed to combine your framework with CSS glow effects on borders.
I figured it should be done using `drop-shadow` as of #7, but I still have the following two problems:

1. how to use the `clip-path` border shape polygon to leave the interior unaffected, i.e. fully transparent and
2. how to control shadow intensity (from the border outwards to customize the glow).

The latter probably does not relate to `augmented-ui`, but rather shows my insufficient CSS expertise. I include it here as I feel there might be related knowledge necessary. Also in my current test case it looks like the `drop-shadow` filter was inherited by all nested elements, but this might also just be an effect occurring due to the missing clipping as of 1. and my specific transparency properties.

I appreciate any hints. Keep it up!

**Comment by James0x57** [comment]
Hello @maxkahnt 

I don't think it's possible to both create an outward glow _and_ clip the middle so there's no glow inside.

BUT

if you just mean you don't want the content to be affected by the drop-shadow, you just have to put your content as a sibling to the decorations.
In this case the decorations will be a container that you apply the change to + an empty augmented element inside as your border. Then make the decoration container + aug child fill the size of the outer element: https://codepen.io/james0x57/pen/BazxRqm

Does that help?

Thank you for checking out augmented-ui!
Please do share the project with me in an issue, comment, or tweet/DM on twitter @James0x57 :)

**Comment by maxkahnt** [comment]
That does help, thank you - also for providing a concise example.

I am not a huge fan of the siblings for semantic reasons but they seem to do the job. I found out that this is something that one would usually do using pseudo elements, but in this case they are occupied already, I suppose ;)

This got me thinking about my original question though: If I move the `drop-shadow` to the class `my-augborder` in your example, it is rendered on the inside of the border. As a consequence I assume that you do not render the border by drawing a seemless polygon whose border is both, the inside and the outside of the visual (blueish) border but rather that you apply some trick (that I do not understand from inspecting the elements) to cut out the interior with other clip-paths in pseudo elements. If this is true, then it might be possible to apply the `drop-shadow` filter to the outermost clip-path'ed pseudo element only and clip-path it with the inner/outer border border (sic!). There might be more element nesting involved to get to work, but I didn't manager to figure it out ad-hoc and wanted to talk back to you first in order to maybe falsify my approach. I cannot yet quite wrap my head around it, because I do not fully understand the clip-path technique you use and it is tedious to inspect on my own with the techniques that I use. If there is a corresponding documentation, I maybe overlooked it. 

P.S. Currently the project is still in a state where I am not confident to make it public, mostly because some of the resources (images) that are used are not verified to be free. But I can provide screenshots for your eyes only soon and the project will eventually be publicly available.

**Comment by James0x57** [comment]
I'm not fully sure what you mean but you're on the right track of thinking for how augmented-ui works. Here's some info that may help:

1. The outer clip-path is applied to the augmented element to trim the exterior away. This also clips any exterior effects like drop-shadow or box-shadow that would normally render outside of the element.

2. Inside of that element is a pseudo element for the border, which clips out the interior instead of the exterior.

3. If the border size was 0px, the two paths would be identical, except the border loops back around the outside so only the exterior is preserved. It is not currently possible in CSS to clip the interior without also clipping the exterior. The exterior in this case is clipped to the normal rectangular shape & size of the element.
  (As you stated, there is no single border polygon, the border is a result of two clip-paths.)

4. For example:
  On the left is an augmented element with an unclipped pseudo element that has a red background.
  On the right is a normal (non-augmented) element with an augmented pseudo element that has a red background, showing the border path only.
     <img src="https://user-images.githubusercontent.com/7545075/97836212-66816300-1ca1-11eb-9520-424ece11dfbc.png" width="500px">

5. You mentioned semantics, which is a fair concern, and fortunately we have a way to indicate that elements are decoration-only by adding the `aria-hidden="true"` attribute to them. This is present in the codepen I linked if you want to see that in use

In short, the code I provided is the minimum needed to draw an augmented border with an exterior glow effect for now. It might be possible to provide a pseudo option in the future that is a standalone border so it's one less element in the markup to glow (drop-shadow), but in either case it would have to be a separate decoration, so there's not much reason to add this as a feature yet.

Does that help clarify things? Anything else I can do to help? :)

**Comment by maxkahnt** [comment]
Thank you again for your extended help.

So you _do_ draw a seemless polygon to fill the border. Using `aug-compile` in chromium the reduced clutter allowed me to come up with the desired effect https://codepen.io/maxkahnt/pen/bGeMMEb. Here is what I did: I took your codepen example above, applied `aug-compile` and added the resulting CSS manually. Copied the element (swapped to relative positioning and added margin). And now I applied your _border magic_ all over again: I cut off the inner glow by adding the clip-path from the `after` pseudo element to the glow element and modified it such that it does not cut the order border strictly (I used search and replace "-1px" -> "-100px" and "+ 1px" => "+ 100px", for anybody copying this: mind the spaces!). Voila!

Now I wonder if I can do this within your css var framework, i.e. before the compile step. I believe that I do need the extra nesting of the `aug-glow` element, so one might be able to work around if one can abuse the inlay. I assume that the explicit extension of the outer clipping could be easily done within your framework by replacing the hard-coded +-1px offsets for the outer rectangle with another css var. My question is how to get hold of the clip-path for the glowing element. A way to do this could be having it as the actual data-augment-ui element and then delegate inlay and border, but I do not know how to actually approach it. Not sure if nested pseudo elements are a thing and can be abused to integrate all this madness into your css framework.

**Comment by James0x57** [comment]
Hah, awesome. I did not realize the clip-path could be extended beyond the background area like that. Nice work

We've got DOM like:
`filter -> aug -> ::aug-border`

And you're copying clip-path from `::aug-border` onto `filter` (plus modifications to extend the outerloop). We obviously can't do that with css since children can't directly influence parent styles and all these vars can't be set on the child and be used on a parent... 
BUT...
If you're okay with the compile step anyway (from the aug-compile issue), we could use js to generate this special path for the `filter` element above:

```js
const augmentedOuterGlowClipPath = augEl => {
  const borderCS = getComputedStyle(augEl, "::after")
  const borderClip = borderCS.getPropertyValue("clip-path")
  const outerSize = "100px"
  const filterClip = borderClip.replace(/-1px ([^,]+,)\s*-1px(?: calc[^,]*,){3} -1px -1px,\s+-1px/, (_, tlBorderIn) => {
    return `-${outerSize} ${tlBorderIn} -${outerSize} calc(100% + ${outerSize}), calc(100% + ${outerSize}) calc(100% + ${outerSize}), calc(100% + ${outerSize}) -${outerSize}, -${outerSize} -${outerSize}, -${outerSize}`
  })
  return filterClip
}
// augEl = ...
augEl.parentNode.style.clipPath = augmentedOuterGlowClipPath(augEl)
```

And here's the original codepen with this function applying the generated path to the filter layer: https://codepen.io/james0x57/pen/LYZmBLx

**Comment by maxkahnt** [comment]
Yep, that works... somewhat. It does no longer render the inlay then if I enable it, though. I did not investigate why that is the case, but it also is lost in your example if you turn it on.

Generally, I'd prefer not having the additional javascript (but I can live with it for optional fanciness).

In the meantime I tried to exploit the mechanism your already implemented to delegate both, border and inlay.
This allows to define the augmentation properties on the parent element, have the glow only applied to the border, and leave the inlay unimpacted. However, I did not yet manage to create the same effect. It might be that I need another nesting level, but if I understand it correctly, the delegate has to be a direct child of the delegating parent. (I am still fiddling with this though, maybe I find some way to make it work.)

**Comment by James0x57** [comment]
You could augment the content at this point and put the inlay there so the filter clip doesn't clip it out: https://codepen.io/james0x57/pen/mdELjoj

**Comment by maxkahnt** [comment]
Yep, that seems fine to me and is a good reason to have siblings :)
I manually did the outer border size replacement in `augmented-ui.min.css` and would love to see a corresponding css variable there.
I will figure out where to place the clip-path injection in my react project now (and edit this comment accordingly).
